### ISSUE TITLE
How to get attachments urls

### ISSUE THREAD

**Issue by juanRabaa** [jgrossi/corcel#427]
- Corcel Version: 2.7.0
- Framework Name & Version: Laravel 5.7
- PHP Version: 7.2.8
- Database Driver & Version: 10.1.34-MariaDB

### Description:
I don't know if there is a better way to get attachments, but right now I'm doing it this way:
`Post::type('attachment')->get();`

This returns all the attachments in the media gallery, but **I only need their urls**.
I tried with `Post::type('attachment')->pluck('url')` but  that throws a sql error

> Unknown column 'url' in 'field list' (SQL: select `url` from `wp_posts` where `post_type` = attachment)

It seems like the 'url' field is generated after the query. It is a combination of 'site-url' + 'wp-content/uploads' + 'wp_attached_file(post meta)'.

Right now, I got it around using the 'guid' field, wich is, unlike the url, in the wp_posts table.
`Post::type('attachment')->pluck('guid')` 
This can generate problems later on, because the guid is not strictly related to the attachment url.

I really dont want to map the collection, so I was wondering if there is a way to do this.

**Comment by jgrossi** [comment]
hi @juanRabaa thanks for sharing 😎about your question, there's no `url` field on `wp_posts` or even a `url` attribute on `Post` or `Attachment` classes. the `guid` column would be the best one here, but I understand your point.

digging into the WP database I saw that some meta are stored for attachments, like the `_wp_attached_file` one, which contains something like this `2011/07/100_5478.jpg`. so I think this is what you're looking for.

the only way I think you can do this with just a single SQL query is mapping the collection and getting the meta value, something like `Post::type('attachment')->with('meta')->get()->map(...)` and creating a new collection of `$post->meta->_wp_attached_file` values, appending your app's URL, for example.

another approach (using 2 SQL queries) would be plucking the post ID and then querying the meta table only for those ids:

```php
$post_ids = Post::type('attachment')->pluck('id')->toArray();
$urls_array = PostMeta::whereIn('post_id', $post_ids)->where('meta_key', '_wp_attached_file')->pluck('meta_value')->toArray();
```

definetely, that's something we can improve in Corcel, creating some easier way to get the attachment's URL, for example. if you have any idea you would like to discuss or even would like to send a PR, yo're welcome. 💪🎉

**Comment by juanRabaa** [comment]
Hey @jgrossi, thanks for the quick response, really apreciate it.

Actually, once you do the `Post::get()`, you get a collection that has an 'url' field. So you can do `Post::get()->pluck('url')` and it will work. 

Somewhere in the `get()` method the url is putted together, so the only way to get this value would be after running `get()` doing `Post::get()->pluck('url')`.

What I wanted to know is if there is a built in way to do this (maybe a query) that doesn't resort to reducing the collection.

I found this article that has a pretty cool query for getting posts urls, but it wont work with post_type = attachment
http://yllus.com/2015/12/08/how-to-get-wordpress-post-permalinks-directly-from-the-mysql-database/

It seems even more easy to get all the urls from attachments because of the `_wp_attached_file` meta, so I made a query that returns all the attachments permalinks.

`
SELECT wpp.ID,
    CONCAT( 
        (SELECT option_value FROM wp_options WHERE option_name='siteurl'), 
        '/wp-content/uploads/',
        wpm.meta_value
    ) AS permalink
FROM 
    wp_posts AS wpp
JOIN wp_postmeta AS wpm
    ON wpm.meta_key = '_wp_attached_file'
    AND wpm.post_id = wpp.ID
WHERE 
    wpp.post_type = 'attachment'
ORDER BY wpp.ID ASC;
`


I think the `/wp-content/uploads/` part in the CONCAT may vary, but I'm not sure.

**Comment by jgrossi** [comment]
@juanRabaa yep, I get your point. actually `url` is an alias to the `guid` if you're using url or guid is the same. Take a look on `Post` where we set all aliases, ~line 99.

that SQL query works for you, and if you really need that at SQL level, sounds good to me. I would go with a collection map call, but this is up to you.

**Comment by juanRabaa** [comment]
@jgrossi so if `url` and `guid ` are the same (in the response from `get`, not in the wp db) there isn't even the need to map the collection, I could just go with my original solution `plunk('guid')`.

Should I be concerned about this values not being the actual url of an attachment file? In case the `guid` column isn't reliable for this, I think it would be useful (and needful) to have something like the `wp_get_attachment_url` function, wich does this concatenation at a php level.

I'm not sure, but I think every other post type would run into this problem as well.

**Comment by jgrossi** [comment]
@juanRabaa yep, a ready-to-use function for that would be great, for sure. like I said, the `url` is just an alias to make it easier to remember. the point is that in WP the hostname can be different if you are using constants, for example, `WP_HOME` so if you're using Laravel I think using `config('app.url')` + `guid` would be better for posts and permalinks, but for files/attachments the WP url is required, because they're stored at WP level.

let's think in something like `$attachment->fullUrl()`. this method would also query for `wp_options` table looking for the home url, like you did before.

if you have another idea, just let me know. let's work on that. thanks!
